provider/qsd.py

```diff
@@ -0,0 +1,446 @@
+        img["format"].update(json.loads(params.get("qsd_image_format", "{}")))
+        img["filter"].update(json.loads(params.get("qsd_image_filter", "{}")))
+
+        img["export"]["id"] = "id_" + name
+        img["export"]["node-name"] = img["filter"]["node-name"] if img["filter"]["driver"] else img["format"][
+            "node-name"]
+        img["export"].update(json.loads(params.get("qsd_image_export", "{}")))
+
+        if img["export"]["type"] == "nbd":
+            # The name is necessary. empty value to simply setting in nbd client
+            img["export"]["name"] = ""
+            addr = json.loads(params.get("qsd_image_export_nbd", '{"type":"unix"'))
+
+            img[addr["type"]].update(addr)
+            if addr["type"] == "unix":
+                if not img[addr["type"]]["path"]:
+                    img[addr["type"]]["path"] = "%s/%s_nbd_%s.sock" % (self.basedir, self.name, name)
+                    # Fix set path
+                    pass
+            img["nbd-server"]["addr"].update(img[addr["type"]])
+
+        elif img["export"]["type"] == "vhost-user-blk":
+            img["unix"]["path"] = "%s/%s_vhost_user_%s.sock" % (self.basedir, self.name, name)
+            img["export"]["addr"] = img["unix"]
+            pass
+        else:
+            raise DeviceError("Unknown export type %s " % img["export"]["type"])
+
+        # Prepare image
+        image_created = False
+
+        if name in params.get("images").split():
+            if params.get("force_create_image") == "yes":
+                image_created = True
+            elif params.get("create_image") == "yes" and not storage.file_exists(img["protocol"]["filename"]):
+                image_created = True
+
+        if image_created:
+            LOG_JOB.info("QSD skip to create image %s ", name)
+        else:
+            LOG_JOB.info("QSD ready to create image %s", name)
+            obj = qemu_storage.QemuImg(params, data_dir.get_data_dir(), name)
+            _, result = obj.create(params)
+            if result.exit_status != 0:
+                raise DeviceError("Failed create image %s " % name)
+            img["image_object"] = obj
+
+        self.images.update({name: img})
+        return img
+
+    def has_option(self, option):
+        """
+        :param option: Desired option
+        :return: Is the desired option supported by current qemu?
+        """
+        return bool(re.search(r"%s" % option, self.__qsd_help,
+                              re.MULTILINE))
+
+    def _probe_capabilities(self):
+        """ Probe capabilities. """
+
+        if self.has_option('--daemonize'):
+            LOG_JOB.info("--daemonize")
+            self.caps.set_flag(Flags.DAEMONIZE)
+        if self.has_option('--pidfile'):
+            LOG_JOB.info("--pidfile")
+            self.caps.set_flag(Flags.PIDFILE)
+
+    def get_pid(self):
+        """ Get QSD pid"""
+        if self.daemon_process:
```

Comment:
self.daemon_process or self.daemon_process() is right?

Reply:
In my understanding ,self.daemon_process it get address of function, self.daemon_process() it get return of function,
self.daemon_process().get_pid()  TypeError: 'Expect' object is not callable

@YongxueHong ,could you please explain more ?

Reply:
> In my understanding ,self.daemon_process it get address of function, self.daemon_process() it get return of function, self.daemon_process().get_pid() TypeError: 'Expect' object is not callable
> 
> @YongxueHong ,could you please explain more ?

Hi @aliang123 

Here the `daemon_process` is the property of the class, it is not a method that is a callable object. but as a property attribute.
```
    @property
    def daemon_process(self):
        """Return daemon process property."""
        return self._daemon_process
```

 You can consider it likes `foo.a` as the following:
```
class Foo(object):
    self.a = "foo"

foo = Foo()
print(foo.a)
```

Reply:
Got it, thanks @YongxueHong